File: leetcode/javascript/leetcode14.js
// 14. Longest Common Prefix

/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function (strs) {
    if (!strs.length) return "";

    let prefix = strs[0];

    for (let i = 1; i < strs.length; i++) {
        console.log("Comparing prefix", prefix, "with", strs[i])
        while (!strs[i].startsWith(prefix)) {
            console.log("Prefix", prefix, "does not start with", strs[i])
            prefix = prefix.slice(0, -1);
            console.log("Updated prefix:", prefix)
            if (prefix === "") return "";
        }
    }

    return prefix;
};

function test(testData, expected) {
    const output = longestCommonPrefix(testData)
    if (JSON.stringify(output) == JSON.stringify(expected)) {
        console.log("Passed:")
    } else {
        console.log("Failed:")
    }
    console.log("Expected", expected)
    console.log("Output", output)
}

test(["flower", "flow", "flight"], "fl")
test(["dog", "racecar", "car"], "")
test(["flow", "flight", "flower",], "fl")
test(["interspecies", "interstellar", "interstate"], "inters")

File: leetcode/python/leetcode14.py
# 14. Longest Common Prefix

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ""

        for i in range(len(strs[0])):
            char = strs[0][i]
            print("char", char, "i", i)
            for s in strs[1:]:
                if i >= len(s) or s[i] != char:
                    return strs[0][:i]

        return strs[0]



def test(test_data, expected):
    sol = Solution()
    output = sol.longestCommonPrefix(test_data)
    if output == expected:
        print("Passed:")
    else:
        print("Failed:")
    print("Expected", expected)
    print("Output", output)


test(["flower","flow","flight"], "fl")
test(["dog","racecar","car"], "")
test(["ab", "a"], "a")